File: src/scalib_ext/geigen/src/geigen.cpp
#include <memory>
#include <iostream>
#include "Eigen/Eigenvalues"
#include "Eigen/Dense"
#include "geigen.h"


using namespace std;
using Eigen::MatrixXd;
using Eigen::VectorXd;
using Eigen::GeneralizedSelfAdjointEigenSolver;
using Eigen::Map;
using Eigen::Stride;
using Eigen::AlignmentType;
using Eigen::Dynamic;

typedef Map<const MatrixXd, AlignmentType::Unaligned, Stride<Dynamic, Dynamic> > MatrixMap;
typedef Spectra::SymGEigsSolver<Spectra::DenseSymMatProd<double>, Spectra::DenseCholesky<double>, Spectra::GEigsMode::Cholesky> GEigenSolverP;

static MatrixMap matrix2map(Matrix matrix);
static Matrix m2m(const MatrixXd& matrix);
static rust::Slice<const double> v2s(const VectorXd& vec);

std::unique_ptr<GEigenSolver> solve_geigen(Matrix a, Matrix b) {
    MatrixMap am = matrix2map(a);
    MatrixMap bm = matrix2map(b);
    GEigenSolver es(am,bm);
    return std::make_unique<GEigenSolver>(es);
}

Matrix get_eigenvecs(const GEigenSolver& solver) {
    return m2m(solver.eigenvectors());
}

rust::Slice<const double> get_eigenvals(const GEigenSolver& solver) {
    return v2s(solver.eigenvalues());
}

std::unique_ptr<GEigenPR> solve_geigenp(Matrix a, Matrix b, uint32_t nev, uint32_t ncv, uint32_t& err) {
    err = 0;
    MatrixMap am = matrix2map(a);
    MatrixMap bm = matrix2map(b);
    Spectra::DenseSymMatProd<double> op(am);
    Spectra::DenseCholesky<double>  Bop(bm);
    // Construct generalized eigen solver object, requesting the largest three generalized eigenvalues
    GEigenSolverP geigs(op, Bop, nev, ncv);

    // Initialize and compute
    geigs.init();
    try {
        geigs.compute(Spectra::SortRule::LargestAlge);
    } catch (const std::runtime_error&) {
        err = 7;
    }
    // Verify Cholesky decomposition
    if (err == 0) {
        if (Bop.info() == Spectra::CompInfo::Successful) {
            err = 0;
        } else if (Bop.info() == Spectra::CompInfo::NotComputed) {
            err = 1;
        } else if (Bop.info() == Spectra::CompInfo::NotConverging) {
            err = 2;
        } else {
            err = 3;
        }
    }

    // Verify Eigendecomposition
    if (err == 0) {
        if (geigs.info() == Spectra::CompInfo::Successful) {
            err = 0;
        } else if (geigs.info() == Spectra::CompInfo::NotComputed) {
            err = 4;
        } else if (geigs.info() == Spectra::CompInfo::NotConverging) {
            err = 5;
        } else {
            err = 6;
        }
    }
    return std::make_unique<GEigenPR>(GEigenPR(geigs.eigenvectors(), geigs.eigenvalues()));
}

Matrix get_eigenvecs_p(const GEigenPR& solver) {
    return m2m(std::get<0>(solver));
}

rust::Slice<const double> get_eigenvals_p(const GEigenPR& solver) {
    return v2s(std::get<1>(solver));
}

static MatrixMap matrix2map(Matrix matrix) {
    Stride<Dynamic, Dynamic> stride(matrix.col_stride, matrix.row_stride);
    MatrixMap map(
            matrix.data,
            matrix.rows,
            matrix.cols,
            stride
            );
    return map;
}

static Matrix m2m(const MatrixXd& matrix) {
    Matrix res;
    res.data= matrix.data();
    res.rows= matrix.rows();
    res.cols= matrix.cols();
    res.row_stride= matrix.rowStride();
    res.col_stride= matrix.colStride();
    return res;
}

static rust::Slice<const double> v2s(const VectorXd& vec) {
    return rust::Slice<const double>(vec.data(), vec.size());
}

File: src/scalib_ext/geigen/src/geigen.rs
//! This module contains the cxx bridge to Eigen.
//! Do not add too many rust functions, as it slows down development (requires to re-compile all
//! the C++ everytime this file is changed).

pub use ffi::*;
use ndarray::ShapeBuilder;
use std::convert::TryInto;

#[cxx::bridge]
mod ffi {
    /// FFI-safe representation of a matrix.
    /// Does not include lifetime as this is not supported by cxx,
    /// but conceptually Matrix<'a> should contain data: &'a [f64].
    /// A matrix is valid is data+(i*row_stride+j*col_stride) points to a valid f64 for lifetime
    /// 'a, for any 0 <= i < rows, 0 <= j < cols.
    struct Matrix {
        data: *const f64,
        rows: isize,
        cols: isize,
        row_stride: isize,
        col_stride: isize,
    }
    unsafe extern "C++" {
        include!("geigen.h");
        /// Full generalized symmetric eigensolver.
        type GEigenSolver;
        // SAFETY: a and b are required to live only for the duration of the call,
        // and a and b must be valid Matrix data.
        unsafe fn solve_geigen(a: Matrix, b: Matrix) -> UniquePtr<GEigenSolver>;
        // Matrix return type implicit lifetime is 'a.
        // SAFETY: consequence of the properties of GEigenSolver
        fn get_eigenvecs<'a>(solver: &'a GEigenSolver) -> Matrix;
        // SAFETY: consequence of the properties of GEigenSolver
        fn get_eigenvals<'a>(solver: &'a GEigenSolver) -> &'a [f64];

        /// Partial generalized symmetric eigensolver.
        type GEigenPR;
        // SAFETY: a and b are required to live only for the duration of the call,
        // and a and b must be valid Matrix data.
        unsafe fn solve_geigenp(
            a: Matrix,
            b: Matrix,
            nev: u32,
            ncv: u32,
            err: &mut u32,
        ) -> UniquePtr<GEigenPR>;
        // Matrix return type implicit lifetime is 'a.
        // SAFETY: consequence of the properties of GEigenSolver
        fn get_eigenvecs_p<'a>(solver: &'a GEigenPR) -> Matrix;
        // SAFETY: consequence of the properties of GEigenSolver
        fn get_eigenvals_p<'a>(solver: &'a GEigenPR) -> &'a [f64];
    }
}

/// Converts a Matrix into a ndarray::ArrayView2<f64>
/// SAFETY: matrix must have lifetime 'a.
pub unsafe fn matrix2array<'a>(
    matrix: Matrix,
) -> Result<ndarray::ArrayView2<'a, f64>, std::num::TryFromIntError> {
    Ok(ndarray::ArrayView::from_shape_ptr(
        (matrix.rows.try_into()?, matrix.cols.try_into()?)
            .strides((matrix.row_stride.try_into()?, matrix.col_stride.try_into()?)),
        matrix.data,
    ))
}

/// Converts a ndarray::ArrayView2<f64> into a Matrix
pub fn array2matrix(a: &ndarray::ArrayView2<f64>) -> Matrix {
    let mut axes = a.axes();
    let ax0 = axes.next().unwrap();
    let ax1 = axes.next().unwrap();
    assert_eq!(ax0.axis.0, 0);
    assert_eq!(ax1.axis.0, 1);
    Matrix {
        data: a.as_ptr(),
        rows: ax0.len.try_into().unwrap(),
        cols: ax1.len.try_into().unwrap(),
        row_stride: ax0.stride,
        col_stride: ax1.stride,
    }
}
